### PR TITLE
Language Server - error locations preparations

### PR DESCRIPTION
**Pull request by fredg1**
I hate having to do this -- that is, having to match the location part as a String, but that's how the full ScriptException is passed right now, so I guess we have to, for now...

### CODE REVIEW

File: src/net/sourceforge/kolmafia/textui/Parser.java [hunk 1]
```diff
@@ -4189,11 +4189,78 @@ private ScriptException parseException(final String expected, final Token found)
...
+          result.append(", ");
+        }
+      }
+
+      if (range.getStart().getCharacter() < range.getEnd().getCharacter()) {
+        result.append("char " + (range.getEnd().getCharacter() + 1));
+      }
+    }
+
+    result.append(")");
```
[the remark this answers is not in the document]
Comment (fredg1) [reply src/net/sourceforge/kolmafia/textui/Parser.java:4261]: Location isn't one of our methods, though...

[the remark this answers is not in the document]
Comment (fredg1) [reply src/net/sourceforge/kolmafia/textui/Parser.java:4261]: ...I guess I can make it a static method instead, though...

File: src/net/sourceforge/kolmafia/textui/Parser.java [hunk 2]
```diff
@@ -4189,11 +4189,78 @@ private ScriptException parseException(final String expected, final Token found)
...
+
+  private ScriptException parseException(final Range range, final String message) {
+    return this.parseException(this.makeLocation(range), message);
+  }
+
+  private ScriptException parseException(final Range start, final Range end, final String message) {
+    return this.parseException(Parser.mergeRanges(start, end), message);
+  }
+
+  private ScriptException parseException(Location location, final String message) {
```
[the remark this answers is not in the document]
Comment (fredg1) [reply src/net/sourceforge/kolmafia/textui/Parser.java:4211]: Is it okay if this is postponed a little?

File: src/net/sourceforge/kolmafia/textui/Parser.java (hunk 1 above)
[the remark this answers is not in the document]
Comment (gausie) [reply src/net/sourceforge/kolmafia/textui/Parser.java]: Ok great. Now add some tests to this function with different ranges sufficient to have 100% coverage on the new function

File: src/net/sourceforge/kolmafia/textui/Parser.java (hunk 2 above)
[the remark this answers is not in the document]
Comment (gausie) [reply src/net/sourceforge/kolmafia/textui/Parser.java:4211]: Yeah sure

File: src/net/sourceforge/kolmafia/textui/Parser.java [hunk 3]
```diff
@@ -4305,6 +4326,54 @@ public static String getLineAndFile(final String fileName, final int lineNumber)
...
   }
 
+  public static final String getFileAndRange(String fileName, final Range range) {
+    if (range == null || Positions.isBefore(range.getEnd(), range.getStart())) {
+      throw new IllegalArgumentException();
+    }
+
+    final StringBuilder result = new StringBuilder();
+
+    result.append("(");
```
Comment (gausie) [inline src/net/sourceforge/kolmafia/textui/Parser.java:4336]: Tiny nitpick but don't add the parens in this function.

**Comment by gausie** [comment]
Before merging this I'd like to see full test coverage of `getFileAndRange`

**Comment by fredg1** [comment]
> Before merging this I'd like to see full test coverage of `getFileAndRange`

I... fail to see what's missing??
firefox_EU3YHwdusL

File: src/net/sourceforge/kolmafia/textui/Parser.java (hunk 3 above)
Comment (fredg1) [reply src/net/sourceforge/kolmafia/textui/Parser.java:4336]: This is just conforming to the behaviour of 
```java
  public static String getLineAndFile(final String fileName, final int lineNumber) {
    if (fileName == null) {
      return "(" + Preferences.getString("commandLineNamespace") + ")";
    }

    return "(" + fileName + ", line " + lineNumber + ")";
  }
```
(which is just above this one)

Comment (gausie) [reply src/net/sourceforge/kolmafia/textui/Parser.java:4336]: I think that's bad behaviour IMO

**Comment by gausie** [comment]
> > Before merging this I'd like to see full test coverage of `getFileAndRange`
> 
> I... fail to see what's missing?? firefox_EU3YHwdusL

Sorry, I don't just mean full coverage incidentally but full coverage via unit testing of that function.

**Comment by gausie** [comment]
Oh I guess you have that actually. Looks fine to me. I still think we should remove the parens, there's no need to keep bad decisions rolling.

**Review by gausie — APPROVED** [review]

**Pull request merged**